python/perf-kernels/fused_moe/moe-gemm.py

```diff
@@ -0,0 +1,495 @@
+        each token.
+    - B: The stacked MOE weight tensor with shape (E, N, K), where E is
+        the number of experts, K is the input feature dimension, and N is
+        the output feature dimension.
+    - C: The output cache tensor with shape (M, topk, N), where M is the
+        total number of tokens post padding, topk is the number of times
+        each token is repeated, and N is the output feature dimension.
+    - sorted_token_ids: A tensor containing the sorted indices of tokens,
+        repeated topk times and arranged by the expert index they are
+        assigned to.
+    - expert_ids: A tensor containing the indices of the expert for each
+        block. It determines which expert matrix from B should be used for
+        each block in A.
+    This kernel performs the multiplication of a token by its corresponding
+    expert matrix as determined by `expert_ids`. The sorting of
+    `sorted_token_ids` by expert index and padding ensures divisibility by
+    BLOCK_SIZE_M, which is necessary to maintain consistency in block matrix
+    multiplication across different blocks processed by the same expert.
+    """
+    pid = tl.program_id(axis=0)
+    num_pid_m = tl.cdiv(EM, BLOCK_SIZE_M)
+    num_pid_n = tl.cdiv(N, BLOCK_SIZE_N)
+    num_pid_in_group = GROUP_SIZE_M * num_pid_n
+    group_id = pid // num_pid_in_group
+    first_pid_m = group_id * GROUP_SIZE_M
+    group_size_m = min(num_pid_m - first_pid_m, GROUP_SIZE_M)
+    pid_m = first_pid_m + ((pid % num_pid_in_group) % group_size_m)
+    pid_n = (pid % num_pid_in_group) // group_size_m
+
+    offs_token_id = pid_m * BLOCK_SIZE_M + tl.arange(0, BLOCK_SIZE_M)
+    offs_token = tl.load(sorted_token_ids_ptr + offs_token_id)
+
+    # Here we assume that valid tokens are in the range [0, M).
+    token_mask = (offs_token >= 0) & (offs_token < EM)
+
+    off_experts = tl.load(expert_ids_ptr + pid_m)
+    offs_bn = (pid_n * BLOCK_SIZE_N + tl.arange(0, BLOCK_SIZE_N).to(tl.int64)) % N
+    offs_k = tl.arange(0, BLOCK_SIZE_K)
+    a_ptrs = A + (offs_token[:, None] // top_k * stride_am + offs_k[None, :] * stride_ak)
+    b_ptrs = B + off_experts * stride_be + (offs_k[:, None] * stride_bk + offs_bn[None, :] * stride_bn)
+
+    accumulator = tl.zeros((BLOCK_SIZE_M, BLOCK_SIZE_N), dtype=tl.float32)
+
+    for k in range(0, tl.cdiv(K, BLOCK_SIZE_K)):
+        # Masking ensures we don't load from invalid tokens or indices
+        a = tl.load(a_ptrs, mask=(token_mask[:, None] & (offs_k[None, :] < K - k * BLOCK_SIZE_K)), other=0.0)
+        b = tl.load(b_ptrs, mask=(offs_k[:, None] < K - k * BLOCK_SIZE_K), other=0.0)
+
+        accumulator = tl.dot(a, b, acc=accumulator)
+        a_ptrs += BLOCK_SIZE_K * stride_ak
+        b_ptrs += BLOCK_SIZE_K * stride_bk
+
+    if MUL_ROUTED_WEIGHT:
+        moe_weight = tl.load(topk_weights_ptr + offs_token, mask=token_mask, other=0)
+        accumulator = accumulator * moe_weight[:, None]
```

Comment:
`topk_weights_ptr` is a 2D tensor of shape M x top_k. `offs_token` is a 1D tensor. Their addition should be a 2D? Don't you need to broadcast offs_token explicitly?
And another question, how does the kernel know this weight belongs to which expert?

Reply:
Yes, there is the broadcasting with [:, None]. Basically one triton block is in charge of only one expert by the expert offset and there are different tokens to handle by the token offsets. and yes, `topk_weights_ptr` has shape (M, top_k).  we load `moe_weight` for different tokens and that's why it has shape (Block_M,). The `accumulator` is with shape (Block_M, Block_N). So we are broadcast along the token dim.

Reply:
> we load moe_weight for different tokens and that's why it has shape (Block_M,)

`topk_weights_ptr` has `top_k` columns, each of which corresponds to one expert. When the kernel load `moe_weight` from `topk_weights_ptr`, which column (expert) is it loading?

Reply:
the `offs_token` contains the sorted (we are sorting by expert idx here) indices of the appearance of the token in the flattened `topk_ids`. For example if we have M = 10 and top_k = 2, the `topk_weights` will have shape (10 , 2) and the flattened `topk_ids` will have shape (20,). Say we look at token 7 and expert 1 (suppose token 7 has expert [0, 1]), it is at `sorted_ids` position 12, with number 16 (7 * 2 + 2). With that as offset, we load the weight according to the token and expert.

I strongly recommend you to read the doc I sent you if you find my explanation confusing. The section `Before Fused MoE: Align tokens to experts`